### ISSUE TITLE
remote debug connect failed

### ISSUE THREAD

**Issue by whitiancai** [openstf/stf#609]
**What is the issue or idea you have?**
i have to run the 'adb connect 192.168.31.202:8088' twice that i can see the remote device after i run 'adb devices',and the connect will occur an error in a few secends

**Does it only happen on a specific device? Please run `adb devices -l` and paste the corresponding row.**
it is not only happen on a specific device,i can control the remote device well before i run 'adb connect 192.168.31.202:8088'. 

**Please provide the steps to reproduce the issue.**
linux ip 192.168.31.202 ,windows ip 192.168.31.135
1.linux run 'stf local --public-ip 192.168.31.202 --allow-remote'
2.windows run 'adb -a -P 5037 fork-server server'
3.linux run 'stf provider --name mikezhoudeMac-mini.local --min-port 7400 --max-port 7700 --connect-sub tcp://127.0.0.1:7114 --connect-push tcp://127.0.0.1:7116 --group-timeout 20000 --public-ip 192.168.31.202 --storage-url http://localhost:7100/ --adb-host 192.168.31.135 --adb-port 5037 --vnc-initial-size 600x800 --allow-remote	'

(i can see the remote device now)

4.see the remote debug cmd in device control page,and  linux run the cmd( 'adb connect 192.168.31.202:7401')
return 'connected to 192.168.31.202:7401'
log:
whitiancai@whitiancai-OptiPlex-3046:~$ adb connect 192.168.31.202:7401
connected to 192.168.31.202:7401

but i can't see any device when i run 'adb devices',and i run the step 4 again and now i can see the remote device
log:
whitiancai@whitiancai-OptiPlex-3046:~$ adb devices
List of devices attached
192.168.31.202:7401	device

5.linux run 'adb logcat',after a few secends the error happened.

**What is the expected behavior?**
the remote device connection will keep alive when i run 'adb connect 192.168.31.202:7401',so i can do something i want 

**Do you see errors or warnings in the `stf local` output? If so, please paste them or the full log here.**
this is the log:
INF/device:plugins:touch 15481 [SODMDIQ499999999] Reading minitouch banner
INF/device:plugins:touch 15481 [SODMDIQ499999999] minitouch says: "Connection established"
WRN/device:plugins:touch 15481 [SODMDIQ499999999] Shell keeping minitouch running ended unexpectedly
INF/device:plugins:touch 15481 [SODMDIQ499999999] Disconnecting from minitouch service
INF/device:plugins:touch 15481 [SODMDIQ499999999] Stopping minitouch service
INF/device:plugins:touch 15481 [SODMDIQ499999999] Launching screen service
INF/device:plugins:touch 15481 [SODMDIQ499999999] Connecting to minitouch service
INF/device:plugins:touch 15481 [SODMDIQ499999999] Reading minitouch banner
INF/device:plugins:touch 15481 [SODMDIQ499999999] minitouch says: "Note: device /dev/input/event4 was outscored by /dev/input/event3 (2100 >= 2000)"
INF/device:plugins:touch 15481 [SODMDIQ499999999] minitouch says: "Type B touch device atmel_mxt_ts_T100_touchscreen (1023x1023 with 10 contacts) detected on /dev/input/event3 (score 2100)"
INF/device:plugins:touch 15481 [SODMDIQ499999999] minitouch says: "Connection established"
WRN/device:plugins:touch 15481 [SODMDIQ499999999] Shell keeping minitouch running ended unexpectedly
INF/device:plugins:touch 15481 [SODMDIQ499999999] Disconnecting from minitouch service
INF/device:plugins:touch 15481 [SODMDIQ499999999] Stopping minitouch service
INF/device:plugins:touch 15481 [SODMDIQ499999999] Launching screen service
INF/device:plugins:touch 15481 [SODMDIQ499999999] Connecting to minitouch service
INF/device:plugins:touch 15481 [SODMDIQ499999999] minitouch says: "Note: device /dev/input/event4 was outscored by /dev/input/event3 (2100 >= 2000)"
INF/device:plugins:touch 15481 [SODMDIQ499999999] minitouch says: "Type B touch device atmel_mxt_ts_T100_touchscreen (1023x1023 with 10 contacts) detected on /dev/input/event3 (score 2100)"
INF/device:plugins:touch 15481 [SODMDIQ499999999] Reading minitouch banner
INF/device:plugins:touch 15481 [SODMDIQ499999999] minitouch says: "Connection established"
WRN/device:plugins:touch 15481 [SODMDIQ499999999] Shell keeping minitouch running ended unexpectedly
INF/device:plugins:touch 15481 [SODMDIQ499999999] Disconnecting from minitouch service
FTL/device:plugins:touch 15481 [SODMDIQ499999999] Touch consumer had an error Error: Failed more than 3 times in 10000ms
    at TouchConsumer._failLimitExceeded (/home/whitiancai/stf/stf-master/lib/units/device/plugins/touch/index.js:230:26)
    at FailCounter.emit (/home/whitiancai/stf/stf-master/node_modules/_eventemitter3@1.2.0@eventemitter3/index.js:117:35)
    at FailCounter.inc (/home/whitiancai/stf/stf-master/lib/util/failcounter.js:29:10)
    at TouchConsumer._outputEnded (/home/whitiancai/stf/stf-master/lib/units/device/plugins/touch/index.js:223:24)
    at RiskyStream.emit (/home/whitiancai/stf/stf-master/node_modules/_eventemitter3@1.2.0@eventemitter3/index.js:115:35)
    at RiskyStream.<anonymous> (/home/whitiancai/stf/stf-master/lib/util/riskystream.js:14:12)
    at emitNone (events.js:91:20)
    at Socket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
FTL/util:lifecycle 15481 [SODMDIQ499999999] Shutting down due to fatal error
INF/provider 15473 [*] Cleaning up device worker "SODMDIQ499999999"
ERR/provider 15473 [*] Device worker "SODMDIQ499999999" died with code 1
INF/provider 15473 [*] Restarting device worker "SODMDIQ499999999"
**Please run `stf doctor` and paste the output here.**
whitiancai@whitiancai-OptiPlex-3046:~$ stf doctor
INF/cli:doctor 16229 [*] OS Arch: x64
INF/cli:doctor 16229 [*] OS Platform: linux
INF/cli:doctor 16229 [*] OS Platform: 4.4.0-75-generic
INF/cli:doctor 16229 [*] Using Node 6.10.0
INF/cli:doctor 16229 [*] Using ZeroMQ 4.2.1
INF/cli:doctor 16229 [*] Using RethinkDB 2.3.5~0trusty
INF/cli:doctor 16229 [*] Using ADB 1.0.39
INF/cli:doctor 16229 [*] Using GraphicsMagick 1.3.25
INF/cli:doctor 16229 [*] Using ProtoBuf 3.3.0

**Comment by sorccu** [comment]
If the connection goes through, you should see that in the logs. It's not
there, so either you should post a longer log, or you're connecting
somewhere else.

The error doesn't have anything to do with remote debugging. Maybe you just
have a device that doesn't work in STF right now. Try a different device.
Make sure that you're able to both see and control the device.

**Comment by whitiancai** [comment]
I am sure i can both see and control the device before i run 'adb connect 192.168.31.202:7041',and it happened on all my devices(tow GALAXY_S4 and one XIAOMI NOTE2) ,not only one deivce.I also restart the stf ,linux pc and windows pc.

Here is the other log when the error occured
INF/triproxy 15333 [app001] PUB socket bound on tcp://127.0.0.1:7111
INF/triproxy 15333 [app001] DEALER socket bound on tcp://127.0.0.1:7112
INF/triproxy 15333 [app001] PULL socket bound on tcp://127.0.0.1:7113
INF/triproxy 15339 [dev001] PUB socket bound on tcp://127.0.0.1:7114
INF/triproxy 15339 [dev001] DEALER socket bound on tcp://127.0.0.1:7115
INF/triproxy 15339 [dev001] PULL socket bound on tcp://127.0.0.1:7116
INF/db 15340 [*] Connecting to 127.0.0.1:28015
INF/processor 15340 [proc001] App dealer connected to "tcp://127.0.0.1:7112"
INF/processor 15340 [proc001] Device dealer connected to "tcp://127.0.0.1:7115"
INF/poorxy 15406 [*] Listening on port 7100
INF/reaper 15352 [reaper001] Subscribing to permanent channel "*ALL"
INF/reaper 15352 [reaper001] Reaping devices with no heartbeat
INF/db 15346 [*] Connecting to 127.0.0.1:28015
INF/db 15352 [reaper001] Connecting to 127.0.0.1:28015
INF/reaper 15352 [reaper001] 1Receiving input from "tcp://127.0.0.1:7111"
INF/reaper 15352 [reaper001] Sending output to "tcp://127.0.0.1:7116"
INF/processor 15346 [proc002] App dealer connected to "tcp://127.0.0.1:7112"
INF/processor 15346 [proc002] Device dealer connected to "tcp://127.0.0.1:7115"
INF/provider 15358 [*] Subscribing to permanent channel "iH0FfHpzR1WQW0jeRoqlHw=="
INF/provider 15358 [*] Sending output to "tcp://127.0.0.1:7116"
INF/provider 15358 [*] Receiving input from "tcp://127.0.0.1:7114"
INF/provider 15358 [*] Tracking devices
INF/storage:plugins:image 15394 [*] Listening on port 7103
INF/storage:plugins:apk 15400 [*] Listening on port 7104
INF/auth-mock 15364 [*] Listening on port 7120
INF/storage:temp 15388 [*] Listening on port 7102
INF/app 15370 [*] Using pre-built resources
INF/app 15370 [*] Listening on port 7105
INF/db 15370 [*] Connecting to 127.0.0.1:28015
INF/websocket 15382 [*] Subscribing to permanent channel "*ALL"
INF/websocket 15382 [*] Listening on port 7110
INF/db 15382 [*] Connecting to 127.0.0.1:28015
INF/websocket 15382 [*] Sending output to "tcp://127.0.0.1:7113"
INF/websocket 15382 [*] Receiving input from "tcp://127.0.0.1:7111"
INF/api 15376 [*] Subscribing to permanent channel "*ALL"
INF/api 15376 [*] Listening on port 7106
INF/api 15376 [*] Sending output to "tcp://127.0.0.1:7113"
INF/api 15376 [*] Receiving input from "tcp://127.0.0.1:7111"
INF/db 15376 [*] Connecting to 127.0.0.1:28015
INF/reaper 15352 [reaper001] Device "SODMDIQ499999999" is present
INF/util:datautil 15382 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.
INF/util:datautil 15376 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.
INF/util:datautil 15376 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.
INF/util:datautil 15376 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.
INF/util:datautil 15376 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.
INF/provider 15358 [*] Found device "192.168.31.202:7401" (offline)
INF/reaper 15352 [reaper001] Device "192.168.31.202:7401" is present
INF/provider 15358 [*] Registered device "192.168.31.202:7401"
INF/provider 15358 [*] Lost device "192.168.31.202:7401" (offline)
INF/provider 15358 [*] Providing all 0 device(s)
INF/provider 15358 [*] Found device "192.168.31.202:7401" (offline)
INF/reaper 15352 [reaper001] Device "192.168.31.202:7401" is present
INF/provider 15358 [*] Registered device "192.168.31.202:7401"
INF/provider 15358 [*] Device "192.168.31.202:7401" is now "device" (was "offline")
INF/device:support:push 15539 [192.168.31.202:7401] Sending output to "tcp://127.0.0.1:7116"
INF/device 15539 [192.168.31.202:7401] Preparing device
INF/device:support:sub 15539 [192.168.31.202:7401] Receiving input from "tcp://127.0.0.1:7114"
INF/device:support:sub 15539 [192.168.31.202:7401] Subscribing to permanent channel "*ALL"
INF/device:support:properties 15539 [192.168.31.202:7401] Loading properties
INF/device:support:sdk 15539 [192.168.31.202:7401] Supports SDK 21
INF/device:support:abi 15539 [192.168.31.202:7401] Supports ABIs arm64-v8a, armeabi-v7a, armeabi
INF/device:resources:minicap 15539 [192.168.31.202:7401] Installing "/home/whitiancai/stf/stf-master/node_modules/minicap-prebuilt/prebuilt/arm64-v8a/bin/minicap" as "/data/local/tmp/minicap"
INF/device:resources:minicap 15539 [192.168.31.202:7401] Installing "/home/whitiancai/stf/stf-master/node_modules/minicap-prebuilt/prebuilt/arm64-v8a/lib/android-21/minicap.so" as "/data/local/tmp/minicap.so"
INF/device:resources:service 15539 [192.168.31.202:7401] Checking whether we need to install STFService
INF/device:resources:service 15539 [192.168.31.202:7401] Running version check
INF/device:resources:service 15539 [192.168.31.202:7401] STFService up to date
INF/device:plugins:service 15539 [192.168.31.202:7401] Launching agent
INF/device:plugins:service 15539 [192.168.31.202:7401] Agent says: "Listening on @stfagent"
INF/device:plugins:service 15539 [192.168.31.202:7401] Agent says: "InputClient started"
INF/device:plugins:service 15539 [192.168.31.202:7401] Launching service
INF/device:plugins:display 15539 [192.168.31.202:7401] Reading display info
INF/device:plugins:phone 15539 [192.168.31.202:7401] Fetching phone info
INF/device:plugins:identity 15539 [192.168.31.202:7401] Solving identity
INF/device:plugins:solo 15539 [192.168.31.202:7401] Subscribing to permanent channel "M3FzTf00pMzxIZuaKcoE5TbzkzA="
INF/device:plugins:screen:stream 15539 [192.168.31.202:7401] Starting WebSocket server on port 7400
FTL/device 15539 [192.168.31.202:7401] Setup had an error Error: listen EADDRINUSE 0.0.0.0:7400
    at Object.exports._errnoException (util.js:1022:11)
    at exports._exceptionWithHostPort (util.js:1045:20)
    at Server._listen2 (net.js:1259:14)
    at listen (net.js:1295:10)
    at net.js:1405:9
    at _combinedTickCallback (internal/process/next_tick.js:77:11)
From previous event:
    at Promise._captureStackTrace (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.11.0@bluebird/js/main/debuggability.js:90:23)
    at Promise._resolveFromResolver (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.11.0@bluebird/js/main/promise.js:480:10)
    at new Promise (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.11.0@bluebird/js/main/promise.js:71:37)
    at createServer (/home/whitiancai/stf/stf-master/lib/units/device/plugins/screen/stream.js:431:14)
    at /home/whitiancai/stf/stf-master/lib/units/device/plugins/screen/stream.js:449:12
    at SerialSyrup.ParallelSyrup.invoke (/home/whitiancai/stf/stf-master/node_modules/_stf-syrup@1.0.0@stf-syrup/lib/parallel.js:54:24)
    at /home/whitiancai/stf/stf-master/node_modules/_stf-syrup@1.0.0@stf-syrup/lib/serial.js:43:33
    at tryCatch1 (/home/whitiancai/stf/stf-master/node_modules/_bluebird@1.1.1@bluebird/js/main/util.js:63:19)
    at Promise$_callHandler [as _callHandler] (/home/whitiancai/stf/stf-master/node_modules/_bluebird@1.1.1@bluebird/js/main/promise.js:708:13)
    at Promise$_settlePromiseFromHandler [as _settlePromiseFromHandler] (/home/whitiancai/stf/stf-master/node_modules/_bluebird@1.1.1@bluebird/js/main/promise.js:724:18)
    at Promise$_settlePromiseAt [as _settlePromiseAt] (/home/whitiancai/stf/stf-master/node_modules/_bluebird@1.1.1@bluebird/js/main/promise.js:896:14)
    at Promise$_fulfillPromises [as _fulfillPromises] (/home/whitiancai/stf/stf-master/node_modules/_bluebird@1.1.1@bluebird/js/main/promise.js:1041:14)
    at Async$_consumeFunctionBuffer [as _consumeFunctionBuffer] (/home/whitiancai/stf/stf-master/node_modules/_bluebird@1.1.1@bluebird/js/main/async.js:74:12)
    at Async$consumeFunctionBuffer (/home/whitiancai/stf/stf-master/node_modules/_bluebird@1.1.1@bluebird/js/main/async.js:37:14)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
FTL/util:lifecycle 15539 [192.168.31.202:7401] Shutting down due to fatal error
INF/provider 15358 [*] Cleaning up device worker "192.168.31.202:7401"
ERR/provider 15358 [*] Device worker "192.168.31.202:7401" died with code 1
INF/provider 15358 [*] Restarting device worker "192.168.31.202:7401"
INF/device:support:push 15554 [192.168.31.202:7401] Sending output to "tcp://127.0.0.1:7116"
INF/device 15554 [192.168.31.202:7401] Preparing device
INF/device:support:sub 15554 [192.168.31.202:7401] Receiving input from "tcp://127.0.0.1:7114"
INF/device:support:sub 15554 [192.168.31.202:7401] Subscribing to permanent channel "*ALL"
INF/device:support:properties 15554 [192.168.31.202:7401] Loading properties
INF/provider 15358 [*] Providing 0 of 1 device(s); waiting for "192.168.31.202:7401"
INF/device:support:sdk 15554 [192.168.31.202:7401] Supports SDK 21
INF/device:support:abi 15554 [192.168.31.202:7401] Supports ABIs arm64-v8a, armeabi-v7a, armeabi
INF/device:resources:minicap 15554 [192.168.31.202:7401] Installing "/home/whitiancai/stf/stf-master/node_modules/minicap-prebuilt/prebuilt/arm64-v8a/bin/minicap" as "/data/local/tmp/minicap"
INF/device:resources:minicap 15554 [192.168.31.202:7401] Installing "/home/whitiancai/stf/stf-master/node_modules/minicap-prebuilt/prebuilt/arm64-v8a/lib/android-21/minicap.so" as "/data/local/tmp/minicap.so"
INF/device:resources:service 15554 [192.168.31.202:7401] Checking whether we need to install STFService
INF/device:resources:service 15554 [192.168.31.202:7401] Running version check
INF/device:resources:service 15554 [192.168.31.202:7401] STFService up to date
INF/device:plugins:service 15554 [192.168.31.202:7401] Launching agent
INF/device:plugins:service 15554 [192.168.31.202:7401] Agent says: "Listening on @stfagent"
INF/device:plugins:service 15554 [192.168.31.202:7401] Launching service
INF/device:plugins:service 15554 [192.168.31.202:7401] Agent says: "InputClient started"
INF/device:plugins:display 15554 [192.168.31.202:7401] Reading display info
INF/device:plugins:phone 15554 [192.168.31.202:7401] Fetching phone info
INF/device:plugins:identity 15554 [192.168.31.202:7401] Solving identity
INF/device:plugins:solo 15554 [192.168.31.202:7401] Subscribing to permanent channel "M3FzTf00pMzxIZuaKcoE5TbzkzA="
INF/device:plugins:screen:stream 15554 [192.168.31.202:7401] Starting WebSocket server on port 7404
INF/provider 15358 [*] Providing 0 of 1 device(s); waiting for "192.168.31.202:7401"
INF/provider 15358 [*] Lost device "192.168.31.202:7401" (device)
IMP/device:plugins:service 15554 [192.168.31.202:7401] Agent connection ended, attempting to relaunch
INF/device:plugins:service 15554 [192.168.31.202:7401] Launching service
IMP/device:plugins:service 15554 [192.168.31.202:7401] Service connection ended, attempting to relaunch
INF/device:plugins:service 15554 [192.168.31.202:7401] Launching service
FTL/device:plugins:service 15554 [192.168.31.202:7401] Agent connection could not be relaunched FailError: Failure: 'device '192.168.31.202:7401' not found'
    at /home/whitiancai/stf/stf-master/node_modules/_adbkit@2.9.2@adbkit/lib/adb/parser.js:183:29
    at tryCatcher (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.9.34@bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.9.34@bluebird/js/main/promise.js:503:31)
    at Promise._settlePromiseAt (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.9.34@bluebird/js/main/promise.js:577:18)
    at Async._drainQueue (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.9.34@bluebird/js/main/async.js:128:12)
    at Async._drainQueues (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.9.34@bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues (/home/whitiancai/stf/stf-master/node_modules/_bluebird@2.9.34@bluebird/js/main/async.js:15:14)
    at runCallback (timers.js:651:20)
    at tryOnImmediate (timers.js:624:5)
    at processImmediate [as _immediateCallback] (timers.js:596:5)
FTL/util:lifecycle 15554 [192.168.31.202:7401] Shutting down due to fatal error
INF/provider 15358 [*] Shutting down device worker "192.168.31.202:7401"
INF/provider 15358 [*] Cleaning up device worker "192.168.31.202:7401"
INF/provider 15358 [*] Gracefully killing device worker "192.168.31.202:7401"
INF/provider 15358 [*] Device worker "192.168.31.202:7401" has retired
INF/provider 15358 [*] Providing all 0 device(s)
INF/util:datautil 15382 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.
INF/util:datautil 15376 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.
INF/util:datautil 15376 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.
INF/util:datautil 15376 [*] Device database does not have an entry for "SODMDIQ499999999" (model "Redmi Note 2"/"hermes"). This is not an error, it only means that we don't have data for it yet.

**Comment by sorccu** [comment]
... are you trying to add a device from another STF to your STF?

**Comment by whitiancai** [comment]
no, i need add the remote device without stf to my stf.

**Comment by sorccu** [comment]
I just read your original description again, and you have two `stf provider` providers running. `stf local` includes `stf provider`. You can not have the same device in two providers or this happens. One provider is trying to set up a device that's already being controlled by another provider, causing the first provider to fail. Then the other provider tries again, and the second provider fails. This continues in a loop until you hit the failure limit.

Basically you're doing it wrong. Maybe you wanted to use `adb tcpip`?

Can't help you with this any more, it's a usage issue.

**Comment by whitiancai** [comment]
i have two `provider` but one is for the localhost and another one is for another pc ,when i run 'stf provider --name mikezhoudeMac-mini.local --min-port 7400 --max-port 7700 --connect-sub tcp://127.0.0.1:7114 --connect-push tcp://127.0.0.1:7116 --group-timeout 20000 --public-ip 192.168.31.202 --storage-url http://localhost:7100/ --adb-host 192.168.31.135 --adb-port 5037 --vnc-initial-size 600x800 --allow-remote	',the second `provider` start for the windows pc ,i can see and control the remote device which from the windows.And then i run 'adb connect 192.168.31.202:7401' twice,you means i register the remote device in my first `provider` in `stf local`?And two `provider` both control the device,so the error happened?
But if i not start the second `provider`,i can't see and control the remote devices.how can i do?please tell me what's the right way to see,control and debug the remote device with the remote pc  not have STF
